File: recall_mobile/lib/models/content_connection.dart
/// Model for AI-discovered connections between content items.
class ConnectionItemSummary {
  final String id;
  final String title;
  final String? summary;
  final String url;
  final String? thumbnailUrl;
  final String? sourceApp;
  final String? category;

  ConnectionItemSummary({
    required this.id,
    required this.title,
    this.summary,
    required this.url,
    this.thumbnailUrl,
    this.sourceApp,
    this.category,
  });

  factory ConnectionItemSummary.fromJson(Map<String, dynamic> json) {
    return ConnectionItemSummary(
      id: json['id'] as String,
      title: json['title'] as String,
      summary: json['summary'] as String?,
      url: json['url'] as String,
      thumbnailUrl: json['thumbnail_url'] as String?,
      sourceApp: json['source_app'] as String?,
      category: json['category'] as String?,
    );
  }
}

class ContentConnection {
  final String id;
  final ConnectionItemSummary sourceItem;
  final ConnectionItemSummary targetItem;
  final double similarityScore;
  final String connectionType;
  final String? aiExplanation;
  final bool isDismissed;
  final DateTime createdAt;

  ContentConnection({
    required this.id,
    required this.sourceItem,
    required this.targetItem,
    required this.similarityScore,
    required this.connectionType,
    this.aiExplanation,
    this.isDismissed = false,
    required this.createdAt,
  });

  factory ContentConnection.fromJson(Map<String, dynamic> json) {
    return ContentConnection(
      id: json['id'] as String,
      sourceItem: ConnectionItemSummary.fromJson(json['source_item'] as Map<String, dynamic>),
      targetItem: ConnectionItemSummary.fromJson(json['target_item'] as Map<String, dynamic>),
      similarityScore: (json['similarity_score'] as num).toDouble(),
      connectionType: json['connection_type'] as String,
      aiExplanation: json['ai_explanation'] as String?,
      isDismissed: json['is_dismissed'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  /// Similarity as a percentage string
  String get similarityPercent => '${(similarityScore * 100).round()}%';
}

File: recall_backend/app/models/content_connection.py
"""
Content connection model.
Represents AI-discovered connections between content items.
"""
from sqlalchemy import Column, String, DateTime, Boolean, Float, Text, ForeignKey, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import uuid
from datetime import datetime
from app.db.session import Base


class ContentConnection(Base):
    """
    Represents a semantic connection between two content items.

    Discovered automatically when new content is saved - the system
    finds similar items in the user's library using vector similarity.
    """
    __tablename__ = "content_connections"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    source_item_id = Column(
        UUID(as_uuid=True),
        ForeignKey("content_items.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    target_item_id = Column(
        UUID(as_uuid=True),
        ForeignKey("content_items.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    similarity_score = Column(Float, nullable=False)
    connection_type = Column(String(50), nullable=False, default='semantic')
    ai_explanation = Column(Text, nullable=True)
    is_dismissed = Column(Boolean, default=False, nullable=False)
    explored_at = Column(DateTime, nullable=True)
    explored_count = Column(Integer, default=0, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    # Relationships
    user = relationship("User", foreign_keys=[user_id])
    source_item = relationship("ContentItem", foreign_keys=[source_item_id])
    target_item = relationship("ContentItem", foreign_keys=[target_item_id])

    def __repr__(self):
        return f"<ContentConnection(id={self.id}, score={self.similarity_score:.2f})>"
